src/profile/ProfileForm.tsx

```diff
@@ -37,585 +25,405 @@ interface ChangedUser {
+
+  const [validateUserNameDebounced] = useDebouncedCallback(
+    validateUsernameAvailable,
+    250
+  );
 
-        if (value.length < 8) {
-          newErrors.push("USER.PASSWORD.TOO_SHORT.8");
-        }
+  const [showPassword, setShowPassword] = useState<boolean>(false);
 
-        if (/^\d*$/.test(value)) {
-          newErrors.push("USER.PASSWORD.ENTIRELY_NUMERIC");
-        }
+  const submit = async (vals: ProfileFormVals) => {
+    const cleaned = cleanFormData(vals);
 
-        dispatch({
-          type: "set data",
-          data: { password: value },
-          errors: { status_code: 0, password: newErrors },
-        });
-        break;
+    const editingMe = props.user?.username === auth.user?.username;
+    try {
+      if (props.user === undefined) {
+        await auth.post<ProfileFormVals>("users/", cleaned);
+        setSuccessMessage(`User ${vals.username} created successfully.`);
+        // refresh the form so we can add a new user
+        props.context.reset();
+      } else {
+        await auth.patch<ProfileFormVals>(
+          editingMe ? "me/" : `users/${vals?.username}/`,
+          cleaned,
+          editingMe
+        );
+        setSuccessMessage(`User ${vals.username} saved successfully.`);
       }
-    }
-  };
-
-  const onSubmit = (evt?: React.FormEvent<HTMLFormElement>) => {
-    evt?.preventDefault();
-
-    if (
-      (Object.keys(state.changedUser) as Array<keyof ChangedUser>).every(
-        (k) =>
-          state.changedUser[k] === undefined ||
-          (k === "is_editor" && state.changedUser[k] === user?.is_editor)
-      )
-    ) {
-      return Promise.resolve();
-    }
-
-    if (Object.values(state.errors).flat().length > 1) {
-      dispatch({
-        type: "set general error",
-        errors: [ERRORS.FORMS.NOT_YET_VALID],
-      });
-      return Promise.reject();
-    }
-
-    dispatch({ type: "is loading" });
-
-    return (user === undefined
-      ? auth.post<ChangedUser>("users/", state.changedUser)
-      : auth.patch<ChangedUser>(
-          user === auth.user ? "me/" : `users/${user?.username}/`,
-          state.changedUser,
-          user === auth.user
-        )
-    ).then(
-      () => {
-        dispatch({ type: "done loading success" });
-      },
-      (err: UserAPIError | string) => {
-        dispatch({ type: "done loading error", errors: err });
+      setGeneralErrors(undefined);
+      if (props.onSubmitSuccessful) {
+        await props.onSubmitSuccessful(vals);
+      }
+    } catch (err) {
+      const apiErrors = err as UserAPIError;
+      props.context.setError(
+        apiErrors.cur_password?.map((error) => ({
+          name: "cur_password",
+          type: "server_error",
+          message: error,
+        })) || []
+      );
+      props.context.setError(
+        apiErrors.password?.map((error) => ({
+          name: "password",
+          type: "server_error",
+          message: error,
+        })) || []
+      );
```

Comment:
Can we make this more general, seeing as server errors can be for any field (and can also be general, with the `detail` field only)? E.g.

```js
for (const [name, errArray] of apiErrors.entries()) {
  if (name === "detail") continue;
  props.context.setError(
    errArray.map((message) => ({
      name,
      type: "server_error",
      message
    })) || []
  );
}
```

Reply:
https://github.com/UWmathNEWS/slugline-web/pull/13/commits/0a4e2cec367de918e069b348696f67eeacd0c269